src/web/core/src/main/java/org/geoserver/web/data/layergroup/LayerGroupProvider.java

```diff
@@ -51,18 +71,33 @@ public LayerGroupProvider(LayerGroupProviderFilter groupFilter) {
     }
 
     @Override
-    protected List<LayerGroupInfo> getItems() {
-        List<LayerGroupInfo> groups = getCatalog().getLayerGroups();
-        if (groupFilter != null) {
-            List<LayerGroupInfo> filtered = new ArrayList<>(groups.size());
-            for (LayerGroupInfo group : groups) {
-                if (groupFilter.accept(group)) {
-                    filtered.add(group);
-                }
-            }
-            groups = filtered;
+    public long size() {
+        return count(getFilter());
+    }
+
+    @Override
+    public int fullSize() {
+        return count(Predicates.acceptAll());
+    }
+
+    @Override
+    public Iterator<LayerGroupInfo> iterator(final long first, final long count) {
+        SortBy sortOrder = getSortOrder();
+        Filter filter = getFilter();
+        try (Stream<LayerGroupInfo> items = query(filter, (int) first, (int) count, sortOrder)) {
+            return items.collect(Collectors.toList()).iterator();
         }
-        return groups;
+    }
+
+    /**
+     * @throws UnsupportedOperationException, this method shouldn't be called at all due to the
+     *     overloading of {@link #size()}, {@link #fullSize()}, and {@link #iterator(long, long)}
```

Comment:
Odd javadoc comment (as in, the description is not relevant to the exception), I'd simplify down to:

```suggestion
     * This method shouldn't be called at all due to the
     *     overloading of {@link #size()}, {@link #fullSize()}, and {@link #iterator(long, long)}
```

(we have no requirement to document exceptions in javadoc)

Reply:
done